File: Leetcode110. Balanced binary tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return dfsHeight(root) != -1;
    }

    public int dfsHeight(TreeNode node) {
        if(node==null)
            return 0;

        // find the heights of the left and right subtree
        // if any of them is -1, don't check the other tree
        int leftHeight = dfsHeight(node.left);
        if(leftHeight==-1)
            return -1;

        int rightHeight = dfsHeight(node.right);
        if(rightHeight==-1)
            return -1;

        if(Math.abs(leftHeight-rightHeight) > 1) {
            return -1;
        }

        return 1 + Math.max(leftHeight, rightHeight);
    }
}

File: Leetcode110. Balanced binary tree/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return (self.dfsHeight(root) != -1)

    def dfsHeight(self, node:TreeNode) -> int:
        if not node:
            return 0

        leftHeight = self.dfsHeight(node.left)
        if(leftHeight==-1):
            return -1

        rightHeight = self.dfsHeight(node.right)
        if(rightHeight==-1):
            return -1

        if(abs(leftHeight - rightHeight) > 1):
            return -1

        return 1 + max(leftHeight, rightHeight)

# runtime - 0ms
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def help(cur,h):
            if not cur: return 0
            
            curL = help(cur.left, h)
            curR = help(cur.right,h)

            if curL == -1 or curR == -1: return -1

            if abs(curL - curR) > 1: return -1
            return 1 + max(curL, curR)

        return True if help(root, 0) != -1 else False

        
